registry/remote/repository.go

```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+	// 1. Validate Content-Type
 	mediaType, _, err := mime.ParseMediaType(resp.Header.Get("Content-Type"))
 	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Content-Type: %w", resp.Request.Method, resp.Request.URL, err)
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: invalid response `Content-Type` header; %w",
+			resp.Request.Method,
+			resp.Request.URL,
+			err,
+		)
 	}
 
+	// 2. Validate Size
 	size := resp.ContentLength
 	if size == -1 {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: unknown response Content-Length", resp.Request.Method, resp.Request.URL)
-	}
-
-	// validate digest if ref is a digest
-	if refDigest, err := ref.Digest(); err == nil {
-		if err = verifyContentDigest(resp, refDigest); err != nil {
-			return ocispec.Descriptor{}, err
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: unknown response Content-Length",
+			resp.Request.Method,
+			resp.Request.URL,
+		)
+	}
+
+	// 3. Validate Digest
+	var dgst, respDigest, refDigest, hdrDigest digest.Digest
+	var refDigestErr, hdrDigestErr error
+	if refDigest, err = ref.Digest(); err != nil {
+		refDigestErr = fmt.Errorf(
+			"%s %q: invalid reference digest: `%s`; %w",
+			resp.Request.Method,
+			resp.Request.URL,
+			respDigest,
+			err,
+		)
+	}
+	if hdrDigestStr := resp.Header.Get("Docker-Content-Digest"); hdrDigestStr != "" {
+		if hdrDigest, err = digest.Parse(hdrDigestStr); err != nil {
+			hdrDigestErr = fmt.Errorf(
+				"%s %q: invalid response header value `Docker-Content-Digest`: `%s`; %w",
+				resp.Request.Method,
+				resp.Request.URL,
+				hdrDigestStr,
+				err,
+			)
 		}
-		return ocispec.Descriptor{
-			MediaType: mediaType,
-			Digest:    refDigest,
-			Size:      size,
-		}, nil
-	}
-
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: empty response Docker-Content-Digest", resp.Request.Method, resp.Request.URL)
 	}
 
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	switch {
+	case calculateDigest:
+		if respDigest, err = calculateDigestFromResponse(resp); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+		}
+		var digests = map[string]digest.Digest{
+			"server-supplied": hdrDigest, // via response header `Docker-Content-Digest`
+			"client-supplied": refDigest, // via request reference string
+		}
+		if err = verifyContentDigests(respDigest, digests); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("digest validation failure; %w", err)
+		}
+		dgst = respDigest
+	case refDigestErr == nil:
+		dgst = refDigest
+	case hdrDigestErr == nil:
+		dgst = hdrDigest
 	}
```

Comment:
There is an edge case not caught by this logic and is a regression: What if a `HEAD` request to `/v2/<repo>/manifests/sha256:xxxxx` does not have a `Docker-Content-Digest` header in the response?

It is expected to have an error returned. However, the code logic will take the client supplied `sha256:xxxxx` as the source of the truth.

Reply:
Thanks for the pointers Shiwei, the first doc is specially useful. I can retrospectively update my commit messages.

As for the second item, I've used the branch name issues/225; trying to do it the [correct way manually](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue), I don't seem to get the gear icon and unable to make the connection.

The third point, I believe I have done that already; can you point me to where I haven't?

And finally the last point, yes I'll keep that in mind; in this case I believe the change is still required; further clarity will no doubt come out at review time.